modules/tensor_mechanics/src/materials/FiniteStrainCPSlipRateRes.C

```diff
@@ -0,0 +1,355 @@
+    FiniteStrainCrystalPlasticity(parameters)
+{
+
+  _resid.resize(_nss,0.0);
+
+  _slip_rate.resize(_nss,0.0);
+  _dsliprate_dgss.resize(_nss,0.0);
+
+  _jacob.resize(_nss);
+  for (unsigned int i = 0;i < _nss; ++i)
+    _jacob[i].resize(_nss,0.0);
+
+  _dgss_dsliprate.resize(_nss);
+  for (unsigned int i = 0;i < _nss; ++i)
+    _dgss_dsliprate[i].resize(_nss,0.0);
+
+  _dsliprate_dsliprate.resize(_nss);
+  for (unsigned int i = 0;i < _nss; ++i)
+    _dsliprate_dsliprate[i].resize(_nss,0.0);
+
+}
+
+void
+FiniteStrainCPSlipRateRes::solveStatevar()
+{
+  preSolveStress();
+  solveStress();
+  if (_err_tol)
+    return;
+  postSolveStress();
+}
+
+void
+FiniteStrainCPSlipRateRes::preSolveStress()
+{
+  FiniteStrainCrystalPlasticity::preSolveStress();
+  for (unsigned int i = 0; i < _nss; ++i)
+    _slip_rate[i] = 0.0;
+}
+
+void
+FiniteStrainCPSlipRateRes::solveStress()
+{
+  Real rnorm, rnorm0, rnorm_prev;
+  unsigned int iter = 0;
+  Real step = 1.0;
+
+#ifdef DEBUG
+  std::vector<Real> rnormst(_maxiter+1),slipratest(_maxiter+1);//Use for Debugging
+#endif
+
+  calcResidJacobSlipRate();
+  if (_err_tol)
+    return;
+  rnorm = calcResidNorm();
+  rnorm0 = rnorm;
+
+#ifdef DEBUG
+  rnormst[iter] = rnorm;
+  Real slipratemax = 0.0;
+  for (unsigned int i = 0; i < _nss; ++i)
+    if ( std::abs(_slip_rate[i]) > slipratemax )
+      slipratemax = std::abs(_slip_rate[i]);
+  slipratest[iter] = slipratemax;
+#endif
+
+  while ( rnorm > _rtol * rnorm0 && rnorm0 > _abs_tol && iter < _maxiter )
+  {
+    calcUpdate();
+
+    for (unsigned int i = 0; i < _nss; ++i)
+      _slip_rate[i] -= _resid[i];
+
+    calcResidualSlipRate();
+    if (_err_tol)
+      return;
+    rnorm_prev = rnorm;
+    rnorm = calcResidNorm();
+
+    if (_use_line_search && rnorm > rnorm_prev && !lineSearchUpdateSlipRate(rnorm_prev, _resid))
+    {
+#ifdef DEBUG
+      mooseWarning("FiniteStrainCrystalPLasticity: Failed with line search");
+#endif
+      _err_tol = true;
+      return;
+    }
+
+    calcJacobianSlipRate();
+    iter++;
+
+#ifdef DEBUG
+    slipratemax = 0.0;
+    for (unsigned int i = 0; i < _nss; ++i)
+      if ( std::abs(_slip_rate[i]) > slipratemax )
+        slipratemax = std::abs(_slip_rate[i]);
+    rnormst[iter] = rnorm;
+    slipratest[iter] = slipratemax;
+#endif
+  }
+
+  if ( iter == _maxiter )
+  {
+#ifdef DEBUG
+    mooseWarning("FiniteStrainCPSlipRateRes: NR exceeds maximum iteration " << iter << " " << rnorm);
+#endif
+    _err_tol = true;
+    return;
+  }
+}
+
+void
+FiniteStrainCPSlipRateRes::calcResidJacobSlipRate()
+{
+  calcResidualSlipRate();
+  if (_err_tol)
+    return;
+  calcJacobianSlipRate();
+}
+
+void
+FiniteStrainCPSlipRateRes::calcResidualSlipRate()
+{
+  RankTwoTensor iden, eqv_slip_incr, ce, ee;
+
+  iden.zero();
+  iden.addIa(1.0);
+
+  for (unsigned int i = 0; i < _nss; ++i)
+    _slip_incr[i] = _slip_rate[i] * _dt;
+
+  eqv_slip_incr.zero();
```

Comment:
The call to zero is never necessary for a freshly constructed RankTwoTensor. Its constructor already initializes the tensor to zero.